test/comma.cpp

```diff
@@ -112,8 +110,8 @@ int test_main(int, char * [])
             (term<int_callable>{{&call_count, &int_called}}(),
              term<void_callable>{{&call_count, &void_called}}());
 
-        BOOST_MPL_ASSERT(
```

Comment:
Most of these changes look good, but why are you having trouble with BOOST_MPL_ASSERT?  It's widely used in Boost tests.  If it's a compilation error, please post the error.  Same if it's a warning too, I guess.

Reply:
```
compile-c-c++ bin.v2\libs\yap\test\comma.test\msvc-14.1\debug\threading-multi\comma.obj
comma.cpp
libs\yap\test\comma.cpp(116): error C2955: 'std::is_same': use of class template requires template argument list
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.15.26726\include\xtr1common(86): note: see declaration of 'std::is_same'

    call "bin.v2\standalone\msvc\msvc-14.1\msvc-setup.bat"  >nul
 cl /Zm800 -nologo @"bin.v2\libs\yap\test\comma.test\msvc-14.1\debug\threading-multi\comma.obj.rsp"

...failed compile-c-c++ bin.v2\libs\yap\test\comma.test\msvc-14.1\debug\threading-multi\comma.obj...
```
IIUC MSVC have some problems with `decltype` in macros.

Reply:
Well that sucks.  Is there any way to fix these tests on MSVC without removing MPL_ASSERT?  MPL_ASSERT is much nicer than static_assert, because when it fails you get the two types being compared printed out in the compiler diagnostics.

Reply:
You can move decltype out of macro and it will help
```cpp
        using type = decltype(evaluate(int_void_expr));
        BOOST_MPL_ASSERT(
            (std::is_same<void, type>));
```

Reply:
Ok, could you reform your PR to do that?  Also, could you rebase the PR on top of my changes from yesterday?  I'd like to see Appveyor (newly added) show green for this PR before accepting.

Reply:
I did. Actually it is not needed to do rebase because travis/appveor makes it automatically (the only problem is lack of start build button on old PRs).

Reply:
Ok, thanks.  It looks like the Appveyor build is still not passing though.

Reply:
What else could it possibly be? :-)

Reply:
For testing whether types are the same, you can also use `BOOST_TEST_TRAIT_TRUE` as in f.ex. https://github.com/boostorg/core/blob/develop/test/is_same_test.cpp#L24.

Reply:
That would work too, if it produces similar diagnostics.  Does it?

Reply:
It's a runtime test, so not quite the same if you'd rather have compile-fail tests. The diagnostic looks like this: https://travis-ci.org/pdimov/variant2/jobs/443239342#L612

Reply:
Ah, got it.  I think I'd rather stick with MPL_ASSERT then.